### PR TITLE
Go: Update the QHelp for `go/command-injection`. 

### PR DESCRIPTION
**Pull request by erik-krogh**
Also fixed some FPs that appeared in the fix-examples in the new QHelp.  

My friend likes this new QHelp. 

[An evaluation](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/go-command__nightly__code-scanning__1/reports) was very uneventful.  
No change in results or performance.

### CODE REVIEW

**Comment by owen-mc** [comment]
@erik-krogh I see that the formatting CI check failed and you fixed it. I worked on it a while ago to make it print out the files that needed to be reformatted, which it hadn't been doing. Did it work for you? Was it easy to see what had failed and what you needed to do?

**Comment by erik-krogh** [comment]
> @erik-krogh I see that the formatting CI check failed and you fixed it. I worked on it a while ago to make it print out the files that needed to be reformatted, which it hadn't been doing. Did it work for you? Was it easy to see what had failed and what you needed to do?

Yes, that was helpful 👍

File: go/ql/src/Security/CWE-078/examples/CommandInjectionGood3.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+func handler(req *http.Request) {
+	repoURL := req.URL.Query()["repoURL"][0]
+	outputPath := "/tmp/repo"
+
+	// Sanitize the repoURL to ensure it does not start with "--"
+	if strings.HasPrefix(repoURL, "--") {
+		log.Fatal("Invalid repository URL")
+		return
+	}
```
Comment (owen-mc) [inline go/ql/src/Security/CWE-078/examples/CommandInjectionGood3.go:18]: Surely you need a command execution guarded by the sanitizer?
```suggestion
	} else {
	  cmd := exec.Command("git", "clone", repoURL, outputPath)
	  err := cmd.Run()
	    	if err != nil {
     		log.Fatal(err)
     	}
  }
```

**Review by owen-mc — COMMENTED** [review]
Generally this looks good. @atorralba should probably do a quick review as well, since I don't have much app sec experience.

**Comment by atorralba** [comment]
> @atorralba should probably do a quick review as well

The `--` sanitizer I like — for argument injection. If I understand it correctly, it's now a sanitizer for the whole command injection query, so would something like this be marked as safe?

```go
func handler(req *http.Request) {
    cmd := req.URL.Query()["cmd"][0]
    if strings.HasPrefix(cmd, "--") {
        log.Fatal("Invalid cmd")
    }
    cmd := exec.Command(cmd)
    cmd.Run()
}
```

If that's the case, this seems dangerous and prone to FNs.

The regex one is a tradeoff: there will be regexes that correctly sanitize the problem, and ones that not. With this sanitizer, we're prioritizing having less FPs in exchange of risking more FNs, which I'm not sure is what we want right now.

Ideally, we'd be able to analyze the regex to ensure that it either:

* Is an allow-list of characters that doesn't include whitespace, or
* Is a disallow-list of characters that includes whitespace.

Of course, this can get very complicated very quickly, and is probably overkill, so I'll leave it up to you if you want to merge as is, improve the sanitizer, or remove it entirely.

Maybe a decent alternative would be accepting less complex solutions that check the presence of whitespaces, like `string.Contains` and the like?

**Comment by erik-krogh** [comment]
> The `--` sanitizer I like — for argument injection. If I understand it correctly, it's now a sanitizer for the whole command injection query, so would something like this be marked as safe?

Yes, that example would spuriously be marked as safe. 
And no, I don't think that's a problem.  
The query doesn't currently use flow-labels, which would allow me to distinguish command-names and command-arguments, so my new sanitizer just sanitizes everything.   

I don't think it's a problem that the sanitizer just sanitizes everything, because the `--` prefix check only makes sense in the places where it actually sanitizes the value.  
The example you showed does not really make sense, because a command-name will never start with `--`, so it wouldn't make sense for that check to be there.    
If you look at [how "--" prefix checks are used in the wild](https://github.com/search?type=code&q=%2Fstrings.HasPrefix%5C%28.*%2C+%22--%22%5C%29%2F+path%3A*.go), you can see that it's used for parsing/sanitizing arguments.  

So yes, FNs are possible, but I think those would be weird. 

> Of course, this can get very complicated very quickly, and is probably overkill, so I'll leave it up to you if you want to merge as is, improve the sanitizer, or remove it entirely. 

We could do something nice, if Go had an implementation of the shared NFA libraries.  
I could do a regular expression that tries to match regular expressions, but I don't think that would be food.  
So I want to keep it as is.  

> Maybe a decent alternative would be accepting less complex solutions that check the presence of whitespaces, like string.Contains and the like? 

Checking just for whitespace is not a valid sanitizer for anything related to command-injection? Can you give me an example of what you mean?

**Comment by owen-mc** [comment]
It's definitely a thing we do elsewhere in the Go library to consider any regexp check a sanitizer without confirming that they've actually checked for the right thing. So I think it's fine to do here too.

**Review by owen-mc — APPROVED** [review]

File: go/ql/src/Security/CWE-078/CommandInjection.qhelp [hunk 2]
```diff
@@ -12,26 +12,49 @@ a malicious user may be able to run commands to exfiltrate data or compromise th
...
 <p>
-If this is not possible, then add sanitization code to verify that the user input is safe before
-using it.
+If given arguments as a single string, avoid simply splitting the string on 
+whitespace. Arguments may contain quoted whitespace, causing them to split into 
+multiple arguments.
+</p>
+<p>
+If this is not possible, sanitize user input to avoid characters like spaces and
+various kinds of quotes that can alter the meaning of the command.
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply go/ql/src/Security/CWE-078/CommandInjection.qhelp:25]: I wanted to mention which characters are dangerous, just to provide some context.  
But deny-list based sanitizers for command injection are generally a bad idea, which is why I didn't do it that way in the example.

**Pull request merged**